airflow/dag_processing/collection.py

```diff
@@ -435,14 +456,17 @@ def add_asset_trigger_references(
         refs_to_add: dict[tuple[str, str], set[str]] = {}
         refs_to_remove: dict[tuple[str, str], set[str]] = {}
         triggers: dict[str, BaseTrigger] = {}
+        active_assets = _find_active_assets(self.assets.keys(), session=session)
```

Comment:
Do we know the DB impact of this? I'm not sure how often the `add_asset_trigger_references` method is called, but it now runs a query with a few joins each time it's called.

Reply:
That's a good question and to be very honest I dont have data so I quite dont know. This is run every-time the dag processor is run.

However, your comment made me realize that an optimization was possible. If no asset is collected by the dag processor, there is no need to run this query.

But if some assets are defined in DAGs, this query will run. If someone has more insight on the impact of this query and whether we should try to optimize it somehow, I am interested :)

Reply:
Interesting, so this is run each time a dag is parsed? That's every 30s right now per dag file. So if you have lots of dag files (even spread across many parsers) that could be a lot of load. But hard to say the scale of things here :face_with_diagonal_mouth:

Reply:
Agree :)

Reply:
With the optimization I just added, DAG file with no asset wont run this query

Reply:
Maybe @dstandish ?

Reply:
I'll proceed with this one. We can always refine the logic and/or the query later if we find out this has some poor consequences on the performances

Reply:
A bunch of queries are run when dag processed. So it’s not really anything new. Though the query does look pretty complicated. Maybe it could be simplified. Should any of those relations have fks that are missing? If the query is expensive it could pcausw problems. If not, then probably non issue.

Reply:
I am not a sqlalchemy expert so I would not be surprised if a better version of this query exists :). Let's look at it.

```
return {
        (asset_model.name, asset_model.uri)
        for asset_model in session.scalars(
            select(AssetModel)
            .join(AssetActive, (AssetActive.name == AssetModel.name) & (AssetActive.uri == AssetModel.uri))
            .where(tuple_(AssetActive.name, AssetActive.uri).in_(name_uri_assets))
            .where(AssetModel.consuming_dags.any(DagScheduleAssetReference.dag_id.in_(active_dags)))
            .options(
                joinedload(AssetModel.consuming_dags).joinedload(DagScheduleAssetReference.dag),
            )
        ).unique()
    }
```

Among a list of assets, I want to retrieve those that are active. By active I mean:
- Exist in DAGs (have not been removed). This is done with the line `.join(AssetActive, (AssetActive.name == AssetModel.name) & (AssetActive.uri == AssetModel.uri))`
- Check only for a given list of assets. This is done with this line ` .where(tuple_(AssetActive.name, AssetActive.uri).in_(name_uri_assets))`. The PK of an asset is a couple (name, uri).
- Assets used to schedule DAGs. This is done with `.where(AssetModel.consuming_dags.any(DagScheduleAssetReference.dag_id.in_(active_dags)))`

I am happy to refactor it if you think there is a way to optimize it. Unfortunately my knowledge of SQLAlchemy is pretty basic so I dont think I'll be able to come up with something better (I was already pleased to have something working :)).

> Should any of those relations have fks that are missing?

Most of them have PK. There is a PK in `AssetActive` on the couple (name, uri). `AssetModel.consuming_dags` is using the magic `relationship` which also uses PKs.

Reply:
One thing I think is helpful is to have sqlalchemy spit out the sql.  then you can look at the generated sql and evaluate, is this optimal?  because we all understand sql better than we understand sqlalchemy. i assume :)

Reply:
just while looking noticed the `unique()`....  `unique()` is... i hate to use the term... a bit of a code smell... when you see `unique()` you know there's a good chance that there is some funnybusiness... probably the query could be rewritten to return the distinct results, and _possibly_ it could be susceptible to join explosion since you _must_ be returning unwanted rows otherwise you would not need to call unique

Reply:
i think the query looks uglier than it is.

it can be simplified to this
```
        select(AssetModel)
        .join(
            AssetActive,
            AssetActive.name == AssetModel.name,
            AssetActive.uri == AssetModel.uri,
        )
        .where(
            tuple_(AssetActive.name, AssetActive.uri).in_(name_uri_assets),
            AssetModel.consuming_dags.any(DagScheduleAssetReference.dag_id.in_(active_dags)),
        )
```
(ignoring the options etc)

Reply:
here's how i obtained that sql printout
```
"""
AIRFLOW__DATABASE__SQL_ALCHEMY_CONN='postgresql://127.0.0.1:25433/airflow?user=postgres&password=airflow'
"""
os.environ["AIRFLOW__DATABASE__SQL_ALCHEMY_CONN"] = (
    "postgresql://127.0.0.1:25433/airflow?user=postgres&password=airflow"
)
os.environ["AIRFLOW__DATABASE__SQL_ALCHEMY_ENGINE_ARGS"] = """{
"echo": true
}"""

import logging

logging.getLogger("sqlalchemy").setLevel(logging.INFO)

from airflow.settings import Session

session = Session()

result = _get_assets(session=session, name_uri_assets=[('a', 'b')], active_dags=["my-dag"])

list(result)
```

Reply:
after refactoring like this
```
    return {
        (asset_model.name, asset_model.uri)
        for asset_model in _get_assets(session, name_uri_assets, active_dags)
    }


def _get_assets(session, name_uri_assets: Iterable[tuple[str, str]], active_dags: Iterable[str]):
    return session.scalars(
        select(AssetModel)
        .join(
            AssetActive,
            and_(
                AssetActive.name == AssetModel.name,
                AssetActive.uri == AssetModel.uri,
            ),
        )
        .where(
            tuple_(AssetActive.name, AssetActive.uri).in_(name_uri_assets),
            AssetModel.consuming_dags.any(DagScheduleAssetReference.dag_id.in_(active_dags)),
        )
        .options(
            joinedload(AssetModel.consuming_dags).joinedload(DagScheduleAssetReference.dag),
        )
    )

```

SO -- hopefully I have helped enable you to look into whether there's an optimal way to query, but i will go back to my day job until summoned again

Reply:
Oh yeah, you did a lot :) Thanks a lot! I'll definitely take it from there :) I think I'll be able to find small improvements :)